pulsar-client/src/main/java/org/apache/pulsar/client/api/ProducerBuilder.java

```diff
@@ -0,0 +1,229 @@
+/**
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.pulsar.client.api;
+
+import java.util.Map;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+import org.apache.pulsar.client.api.PulsarClientException.ProducerQueueIsFullError;
+
+public interface ProducerBuilder {
+
+    Producer create() throws PulsarClientException;
```

Comment:
nit: since this is a builder, do we want to go with the canonical 'build' terminology instead?

Reply:
Yes, I was debating about this. My reasoning here was to differentiate the actions: 

```java
client.newProducer().topic(MY_TOPIC).create();
// versus
client.newConsumer().topic(MY_TOPIC).subscriptionName(MY_SUB).subscribe();
```

Subscribe here implies to create a durable `Subscription` on the topic, not just building an ephemeral `Consumer` instance.

I'm not 100% settled on this. Please anyone share opinions ;)